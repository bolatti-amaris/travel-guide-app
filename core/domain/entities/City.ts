export interface City {
    id: number;
    key: string;
    name: string;
    nativeName: string;
    currency: string;
    language: string;
  }