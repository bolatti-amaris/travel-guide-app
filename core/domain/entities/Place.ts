export interface Place {
  type: 'monument' | 'restaurant';
  name: string;
  coordinates: [number, number];
}
