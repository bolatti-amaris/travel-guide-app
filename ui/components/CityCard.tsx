import { View, Text } from 'react-native'
import React from 'react'

interface CityCardProps {
    id: number,
    key: string,
    name: string,
    nativeName: string,
    currency: string,
    language: string,
}

const CityCard = (city : CityCardProps) => {
  return (
    <View>
      <Text>{ city.name }</Text>
    </View>
  )
}

export default CityCard