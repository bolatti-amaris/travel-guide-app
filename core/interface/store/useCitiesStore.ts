import { create } from "zustand";
import { devtools, persist, createJSONStorage } from "zustand/middleware";
import AsyncStorage from "@react-native-async-storage/async-storage";

import { City } from "@/core/domain/entities/City";
import { GetAllCities } from "@/core/domain/usecases/GetAllCities";
import { CityRepositoryImpl } from "@/core/infrastructure/repositories/CityRepositoryImpl";

interface CitiesState {
  cities: City[];
  selectedCity: City | null;
  isLoading: boolean;
  error: string | null;
  fetchCities: () => Promise<void>;
  setSelectedCity: (city: City) => void;
  reset: () => void;
}

export const useCitiesStore = create<CitiesState>()(
  devtools(
    persist(
      (set, get) => {
        const repository = new CityRepositoryImpl();
        const getAllCities = new GetAllCities(repository);

        return {
          cities: [],
          selectedCity: null,
          isLoading: false,
          error: null,

          fetchCities: async () => {
            set({ isLoading: true, error: null });

            try {
              const cities = await getAllCities.execute();
              set({ cities, isLoading: false });
              await AsyncStorage.setItem("cities", JSON.stringify(cities));
            } catch (err: any) {
              let message = "Unexpected error";
              if (err.message === "No internet connection") {
                message =
                  "No internet connection. Please check your network and try again.";
              } else if (err.message.toLowerCase().includes("timeout")) {
                message = "The request timed out. Please try again later.";
              } else if (err.message.toLowerCase().includes("graphql")) {
                message =
                  "There was a problem retrieving city data from the server.";
              } else if (err.message) {
                message = err.message;
              }

              const storedCities = await AsyncStorage.getItem("cities");
              if (storedCities) {
                set({ cities: JSON.parse(storedCities), isLoading: false });
              }

              set({
                error: message,
                isLoading: false,
              });
            }
          },

          setSelectedCity: (city) => {
            set({ selectedCity: city });
          },

          reset: () => {
            set(
              {
                cities: [],
                selectedCity: null,
                error: null,
                isLoading: false,
              },
              false,
              "cities/reset",
            );
          },
        };
      },
      {
        name: "cities-storage",
        storage: createJSONStorage(() => AsyncStorage),
      },
    ),
    { name: "CitiesStore" },
  ),
);
