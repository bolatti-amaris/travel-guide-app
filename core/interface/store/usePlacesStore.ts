import { create } from 'zustand';
import { devtools } from 'zustand/middleware';
import { Place } from '@/core/domain/entities/Place';
import { GetPlacesByCityKey } from '@/core/domain/usecases/GetPlacesByCityKey';
import { PlaceRepositoryImpl } from '@/core/infrastructure/repositories/PlaceRepositoryImpl';

interface PlacesState {
  places: Place[];
  isLoading: boolean;
  error: string | null;
  fetchPlaces: (key: string) => Promise<void>;
  reset: () => void;
}

export const usePlacesStore = create<PlacesState>()(
  devtools(
    (set) => {
      const repository = new PlaceRepositoryImpl();
      const getPlacesByCityKey = new GetPlacesByCityKey(repository);

      return {
        places: [],
        isLoading: false,
        error: null,

        fetchPlaces: async (key: string) => {
          set({ isLoading: true, error: null }, false, 'places/fetchPlaces');
          try {
            const places = await getPlacesByCityKey.execute(key);
            set(
              { places, isLoading: false },
              false,
              'places/fetchPlacesSuccess',
            );
          } catch (err: any) {
            set(
              {
                error: err.message || 'Error fetching places',
                isLoading: false,
              },
              false,
              'places/fetchPlacesError',
            );
          }
        },

        reset: () => {
          set(
            { places: [], error: null, isLoading: false },
            false,
            'places/reset',
          );
        },
      };
    },
    { name: 'PlacesStore' },
  ),
);
