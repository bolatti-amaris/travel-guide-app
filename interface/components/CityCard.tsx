import { City } from '@/core/domain/entities/City';
import { useCitiesStore } from '@/core/interface/store/useCitiesStore';
import { useRouter } from 'expo-router';
import React from 'react';
import { View, Text, Pressable } from 'react-native';
import { Icon } from '../shared/Icon';

interface CityCardProps {
  city: City;
}

const CityCard = ({ city }: CityCardProps) => {
  const router = useRouter();
  const setSelectedCity = useCitiesStore((state) => state.setSelectedCity);
  const handlePress = () => {
    setSelectedCity(city);
    router.push('/cities/cityDetails');
  };

  return (
<Pressable onPress={handlePress} className="bg-slate-50 p-4 rounded-sm mb-3">
  <View className="flex-row justify-between items-center">
    <View>
      <View className="flex-row flex-wrap items-baseline mb-1">
        <Text className="text-2xl font-work-bold text-gray-800 mr-2">{city.name}</Text>
        <Text className="text-xl font-work-medium text-gray-400 italic">({city.nativeName})</Text>
      </View>
      <View className="flex-row flex-wrap">
        <Text className="text-base text-gray-700 font-work-light mr-2">Currency: {city.currency}</Text>
        <Text className="text-base text-gray-700 font-work-light">Language: {city.language}</Text>
      </View>
    </View>

    <Icon name="ChevronRight" color="#9ca3af" />
  </View>
</Pressable>

  );
};

export default CityCard;