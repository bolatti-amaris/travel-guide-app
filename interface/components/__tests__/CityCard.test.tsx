import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import CityCard from "../CityCard";
import { useRouter } from "expo-router";
import { useCitiesStore } from "@/core/interface/store/useCitiesStore";
import { City } from "@/core/domain/entities/City";

jest.mock("expo-router", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@/core/interface/store/useCitiesStore", () => ({
  useCitiesStore: jest.fn(),
}));

const mockCity: City = {
  id: 1,
  key: "paris",
  name: "Paris",
  nativeName: "Paris",
  currency: "EUR",
  language: "fr",
};

const mockPush = jest.fn();
const mockSetSelectedCity = jest.fn();

describe("CityCard", () => {
  const mockedUseCitiesStore = useCitiesStore as unknown as jest.Mock;

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({ push: mockPush });

    mockedUseCitiesStore.mockImplementation((selector) =>
      selector({ setSelectedCity: mockSetSelectedCity }),
    );

    jest.clearAllMocks();
  });

  describe("render", () => {
    it("muestra la información de la ciudad correctamente", () => {
      const { getByTestId } = render(<CityCard city={mockCity} />);

      expect(getByTestId("city-name").props.children).toBe("Paris");
      expect(getByTestId("city-native-name").props.children).toContain("Paris");
      expect(getByTestId("city-currency").props.children).toContain("EUR");
      expect(getByTestId("city-language").props.children).toContain("fr");
    });

    it("aplica correctamente las clases Tailwind", () => {
      const { getByTestId } = render(<CityCard city={mockCity} />);

      expect(getByTestId("city-card").props.className).toContain(
        "bg-card-background",
      );
      expect(getByTestId("city-card").props.className).toContain(
        "dark:bg-card-background-dark",
      );
      expect(getByTestId("city-card").props.className).toContain("p-4");

      expect(getByTestId("city-name").props.className).toContain("text-2xl");
      expect(getByTestId("city-name").props.className).toContain(
        "font-work-bold",
      );
      expect(getByTestId("city-name").props.className).toContain(
        "dark:text-text-header-dark",
      );

      expect(getByTestId("city-native-name").props.className).toContain(
        "text-xl",
      );
      expect(getByTestId("city-native-name").props.className).toContain(
        "italic",
      );
      expect(getByTestId("city-native-name").props.className).toContain(
        "dark:text-text-dark",
      );

      expect(getByTestId("city-currency").props.className).toContain(
        "font-work-light",
      );
      expect(getByTestId("city-currency").props.className).toContain(
        "text-text",
      );

      expect(getByTestId("city-language").props.className).toContain(
        "dark:text-text-dark",
      );
    });

    it("coincide con el snapshot", () => {
      const tree = render(<CityCard city={mockCity} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("interacción", () => {
    it("navega y actualiza el store al pulsar", () => {
      const { getByTestId } = render(<CityCard city={mockCity} />);
      fireEvent.press(getByTestId("city-card"));

      expect(mockSetSelectedCity).toHaveBeenCalledWith(mockCity);
      expect(mockPush).toHaveBeenCalledWith("/cities/cityDetails");
    });
  });
});
