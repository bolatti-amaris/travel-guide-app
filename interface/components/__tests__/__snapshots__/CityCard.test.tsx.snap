// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityCard render coincide con el snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  className="bg-card-background dark:bg-card-background-dark p-4 rounded-sm mb-3"
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="city-card"
>
  <View
    className="flex-row justify-between items-center"
  >
    <View>
      <View
        className="flex-row flex-wrap items-baseline mb-1"
        testID="city-names"
      >
        <Text
          className="text-2xl font-work-bold text-text-header dark:text-text-header-dark mr-2"
          testID="city-name"
        >
          Paris
        </Text>
        <Text
          className="text-xl font-work-medium text-text dark:text-text-dark italic"
          testID="city-native-name"
        >
          (
          Paris
          )
        </Text>
      </View>
      <View
        className="flex-row flex-wrap"
        testID="city-details"
      >
        <Text
          className="text-base text-text dark:text-text-dark font-work-light mr-2"
          testID="city-currency"
        >
          Currency: 
          EUR
        </Text>
        <Text
          className="text-base text-text dark:text-text-dark font-work-light"
          testID="city-language"
        >
          Language: 
          fr
        </Text>
      </View>
    </View>
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={24}
      bbWidth={24}
      fill="none"
      focusable={false}
      height={24}
      meetOrSlice={0}
      minX={0}
      minY={0}
      stroke="#262626"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth={2}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 24,
            "width": 24,
          },
        ]
      }
      vbHeight={24}
      vbWidth={24}
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "payload": 4280690214,
            "type": 0,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
      >
        <RNSVGPath
          d="m9 18 6-6-6-6"
          fill={null}
          propList={
            [
              "fill",
              "stroke",
              "strokeWidth",
              "strokeLinecap",
              "strokeLinejoin",
            ]
          }
          stroke={
            {
              "payload": 4280690214,
              "type": 0,
            }
          }
          strokeLinecap={1}
          strokeLinejoin={1}
          strokeWidth={2}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
